### ISSUE TITLE
Home Assistant - Entities stop reporting after a while (Ubuntu)

### ISSUE THREAD

**Issue by Anto79-ops** [timmo001/system-bridge#2366]
### Description

Hello!

Just wanted to mention an ~~couple of~~  issue when using System Bridge on Ubuntu and Home Assistant 2023.4.

1. ~~It seems the integration blocks HA from loading at startup. I have to disable System Bridge before a restart, and then enable it again once HA has started. I've reported the issue on the HA core issue [here](https://github.com/home-assistant/core/issues/90523)~~ this is now fixed in 2023.4.5. Thank you. 

2. The next issue I see is that I'm finally seeing that entities are being reported into HA, and closed this issue [here](https://github.com/home-assistant/core/issues/77731). However, after a while, it seems the entities stop reporting. For example, here is CPU usage sensor after restarting Ubuntu:

![image](https://user-images.githubusercontent.com/79593761/230446264-cc74c9cd-c9b5-4272-ab11-c7587d77fdfc.png)

After about 6 hrs, the entities stop reporting.

When checking if the service is still running in Ubuntu, it seems to be running fine:

![image](https://user-images.githubusercontent.com/79593761/230448889-4d7dfeff-8d36-46ea-a36a-5bbc17626709.png)

When logging into system bridge via the URL on Ubuntu, it is still running, I think?

![image](https://user-images.githubusercontent.com/79593761/230447197-98bb2c2e-1b37-461e-966c-115123ce5fe5.png)

How can I tell if the local url is updating?  Are they supposed to update live? as the numbers are a bit cryptic. This would at least tell if its an HA issue or systembridge backend issue...I think

Any help?

### What Platform / OS are you running?

Linux

### What version are you running?

3.6.2

**Comment by Anto79-ops** [comment]
Unfortunately, still a problem.

**Comment by Anto79-ops** [comment]
Unfortunately, still a problem for me. thanks.

**Comment by 007trains** [comment]
I'm experiencing the same issue on version 3.6.2:

Exactly 6 hours after starting systembridge stops reporting all cpu and memory entities
Disk usage entities haven't been effected

**Comment by Anto79-ops** [comment]
Just wanted to mention that this is still happening on 3.8.1.

@007trains you mentioned that disk entities still work? That is interesting as  I don't think they work on mine but I'll check again.

**Comment by shaunsetlock** [comment]
I have experienced this issue, as well. HA stops recording data after several hours, but systembridge appears to be running happily. Restarting systembridge (in my case, a systemd service) seems to kick things back into working.

- HA Core 2023.7.3
- HA OS 10.3
- All machines are running as a systemd service on Linux. 
- There are both bare metal installs and Linux VM's on Proxmox. 
- Operating systems are Debian 12 or Ubuntu 22.04.

Great project, thank you Tim.

**Comment by funforums** [comment]
Hi, 

First of all, thanks for this cool project, it allows bringing things to the next level here at home.

I just wanted to report on my experience with this problem.
I am running System Bridge on a Windows tablet and I could replicate a very similar behaviour - what gets it back working for me is restarting either the Home Assistant integration or the tablet. My guess looking at the bridge logs, when it hangs, it seems to be updating sensor data but never sending the data to Home Assistant.

Is there some kind of "recall" that happens to the device when the system bridge integration has been reload on Home Assistant ? Maybe the effect of the reload of the HASSIO integration wakes the system bridge process up somehow ?

**Comment by funforums** [comment]
Something more to add. Systembridge's HTTP server seems still to respond - but it seems it reports always the same values.
Home assistant reports that the only thing that is changing is the timestamp to the battery discharge, which obviously will change forever since the time moves forward and the devices is always plugged in.
It seems to me that the reason needs to be found in the code that updates all the sensors in general - all the values seem to never change from the perspective of Home assistant until I restart the device or systembridge (since access to this device is kinda weird, I always restart the device, but I assume restarting system bridge will work too).

**Comment by funforums** [comment]
So I have started to look into it and it seems to me that python is hanging on the `asyncio.sleep(120)` call [here](https://github.com/timmo001/system-bridge/blob/c8888cca3ad30d5ee03426da0e552c05f900dea7/backend/systembridgebackend/server/__init__.py#L229), based on the analysis of my device's logs and the printouts made in the code. Now I am not a python expert, but a quick google search returned cases in which calls to `asyncio.sleep` seemed to hang. Any suggestion what could be going on?

**Comment by Anto79-ops** [comment]
thanks, hopefully Tim can comment.

**Comment by funforums** [comment]
Maybe I can tag him @timmo001 . In the meantime, since it may have to do with `asyncio`, may I ask you @Anto79-ops what python version are you running system bridge with? Mine is `3.11.4`

**Comment by Anto79-ops** [comment]
I can check, but when I run my service starts, it uses python 3.10, for example

 `python3.10 -m pip install --upgrade systembridgeshared systembridgebackend systembridgecli`

**Comment by funforums** [comment]
A quick update: I have now spent some time to upgrade my device environment to Python 3.12 and System Bridge 3.9.0. Will report on how it goes in the next few days, hopefully I won't be able to reproduce it anymore.

**Comment by Anto79-ops** [comment]
just curious, how you updated to 3.12 on Ubuntu. If you followed a guide that would be great. Last time I updated from 3.9 to 3.10 I completely destroyed my system and had to start from scratch. I think the problem was I tried to uninstall 3.9 after installing 3.10 which was a mistake.

**Comment by funforums** [comment]
> just curious, how you updated to 3.12 on Ubuntu. If you followed a guide that would be great. Last time I updated from 3.9 to 3.10 I completely destroyed my system and had to start from scratch. I think the problem was I tried to uninstall 3.9 after installing 3.10 which was a mistake.

My device is running Windows, there the replacement is quite easy - but I suspect you´d need to identify which python package you need to install and do it in the reverse order - first take the older python out, and then bring the new one in. Some distributions however should also allow you to install multiple python versions at the same time, at least in the context of python2 and python3 ....

**Comment by funforums** [comment]
I have replicated the issue in 3.12 too sadly. @Anto79-ops did you experience this problem also with Python 3.9 ?

**Comment by Anto79-ops** [comment]
Ahh OK, I was about to update to 3.12. Thanks for sharing @funforums 

I only ever had 3.10, as system bridge I believe requires 3.10 at minimum 

Hopefully some information above could shed some light on this issue.

**Comment by funforums** [comment]
> Ahh OK, I was about to update to 3.12. Thanks for sharing @funforums
> 
> I only ever had 3.10, as system bridge I believe requires 3.10 at minimum
> 
> Hopefully some information above could shed some light on this issue.

No, it should work with 3.9 as well. Apparently there was some change in the way Python performed asyncio calls with 3.10. 
In the meantime, I have made a code change locally to see if the problem can be worked around by using asyncio differently. Since I need to wait an average of 3-4 days to see if it is at least better, it will take time for me to report back.

**Comment by funforums** [comment]
Sadly my attempt failed - the problem still occurs. Still investigating.

**Comment by funforums** [comment]
A short update: My new hypotesis is that the thread stops working because of too many recursive calls to `update_data`. I have made another change to test this and so far so good, but it is still too early to call. I should be able to see a noticeable difference if it does not crash in the next few days (usually my average failure rate one every 1.5 - 2 days) and now I am above 1day of runtime. Will keep you posted.

**Comment by Anto79-ops** [comment]
thanks. Do you mind sharing what you changed? perhaps I can change the same on my end if its not too complicated.

**Comment by funforums** [comment]
> thanks. Do you mind sharing what you changed? perhaps I can change the same on my end if its not too complicated.

Hey, I made a Pull request with what I believe are the essence of the fix after some cleanup. @timmo001 @Anto79-ops here are the changes: https://github.com/timmo001/system-bridge-backend/pull/28
I am still testing after this cleanup, but for me it seemed to run over 4 days without interrupting, which is already a lot more than usual. I believe the reason of the bug was that the previous version of the code was hitting the recursion limit of the python distribution, which is probably different across OS and distributions or single installations. I will be continuing to test this, maybe you could test this too @Anto79-ops ?

**Comment by Anto79-ops** [comment]
oh thanks, @funforums I can implement these changes and it seems its less 10 lines you changed.

So, i I installed system bridge via pip, where can I find this file to sudo nano it? thanks

**Comment by funforums** [comment]
> oh thanks, @funforums I can implement these changes and it seems its less 10 lines you changed.
> 
> So, i I installed system bridge via pip, where can I find this file to sudo nano it? thanks

I wouldn't know for sure, but maybe you can find out by `pip show systembridgebackend`.

**Comment by funforums** [comment]
> oh thanks, @funforums I can implement these changes and it seems its less 10 lines you changed.
> 
> So, i I installed system bridge via pip, where can I find this file to sudo nano it? thanks

One more thing @Anto79-ops : if you have a script that uodates system bridge packages before starting it, you should disable this routine - you don't want updates to overwrite the changes you do manually.

**Comment by Anto79-ops** [comment]
ok thanks @funforums 

it was `pip show systembridgebackend`

I made the changes manually and stopped and started the service and its running now. If it makes its past 10 hrs, it would already be a success.

I don't think I have the auto-update script enabled (unless its be default), I wouldn't even know how to enable it!

keep you posted

**Comment by funforums** [comment]
> ok thanks @funforums
> 
> it was `pip show systembridgebackend`
> 
> I made the changes manually and stopped and started the service and its running now. If it makes its past 10 hrs, it would already be a success.
> 
> I don't think I have the auto-update script enabled (unless its be default), I wouldn't even know how to enable it!
> 
> keep you posted

I am referring to the way you said you start system-bridge. Didnt you say you run `pip install --upgrade` every time you start your service? That might overwrite your changes.

**Comment by Anto79-ops** [comment]
Oh I see the confusion now, I only use 

`pip install --upgrade...` to run updates but to run the program I use a service as shown in the original post of the issue, above.

@funforums @timmo001 i have some good news! With the changes suggested above, my system bridge has been running for almost 18 hrs now which is about 8 to 10hrs more than it would be running without the changes to the file.

This is the longest SystemBridge has been running, ever, for me.

Thanks @funforums for finding a solution.

**Comment by Anto79-ops** [comment]
ok the believe the problem is solved, HOWEVER, i did end up reverting the changes because having SystemBridge run that long (i.e. 24/7), exploited another existing problem which was a bit more serious.

Whenever SystemBridge runs, my HDD gets thrashed its constantly writing something. Here's a Glances while the SystemBridge service is running:

![image](https://github.com/timmo001/system-bridge/assets/79593761/060ab682-9e6e-4e2e-9ef2-9049c36090dd)

and this is after I stopped the service:

![image](https://github.com/timmo001/system-bridge/assets/79593761/d9e922d3-8784-443f-b3b2-bbce4e371ae1)

This is not a new issue, it was always there, however, having my HDD at 66% activity (according `atop`) for 24/7 is not ideal. At least with the old version, my system will update every time I restart ubuntu and the sensors will update then for 6 hrs.

Not sure why there is so much disk activity with SystemBridge.
